### PR TITLE
propagates csi feature states from supervisor cluster to workload clusters

### LINKED ISSUE #2568: VsphereCSIConfigController should sync featurestate configs from supervisor cluster to workload cluster
(This is used to request new product features)

**Describe the feature request**
CSI feature state ConfigMap 'vmware-system-csi/csi-feature-states' should be synced from supervisor to workload cluster. 
**Describe alternatives you've considered**
N/A
**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [x] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by peterochodo**
### What this PR does / why we need it
This reconciles the ConfigMap 'vmware-system-csi/csi-feature-states' that is present in the supervisor cluster of tkgs, by propagating its configuration to the value.yaml file used by vsphere-pv-csi addon in workload clusters

### Which issue(s) this PR fixes

Fixes #2568 

### Describe testing done for PR
- env tests

### Release note

```release-note
package-based-lcm: Sync csi feature states from supervisor cluster to workload cluster
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/test/testutil/test_helpers.go [hunk 1]
```diff
@@ -155,6 +155,33 @@ func DeleteResources(f *os.File, cfg *rest.Config, dynamicClient dynamic.Interfa
...
 }
 
+// EnsureResources verifies that resources exist, creating it if necessary
+func EnsureResources(f *os.File, cfg *rest.Config, dynamicClient dynamic.Interface) error {
+	decoder, mapper, err := parseObjects(f, cfg)
+	if err != nil {
+		return err
+	}
+
+	for {
```
[the remark this answers is not in the document]
Comment (peterochodo) [reply addons/test/testutil/test_helpers.go:165]: very unlikely. The function 'decoder.Decode(&rawObj)' invoked by 'getResource(...)', is bound to ultimately return an 'io.EOF' error, or some other kind of error.

File: addons/controllers/csi/vspherecsiconfig_utils.go [hunk 2]
```diff
@@ -55,6 +55,20 @@ func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesParavirtual(
 	dvs.VSpherePVCSI.Namespace = "vmware-system-csi"
```
[the remark this answers is not in the document]
Comment (peterochodo) [reply addons/controllers/csi/vspherecsiconfig_utils.go]: updated

File: addons/controllers/csi/vspherecsiconfig_controller.go [hunk 3]
```diff
@@ -39,10 +43,46 @@ type VSphereCSIConfigReconciler struct {
...
+
+	fsPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return isFeatureStatesConfigMap(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return isFeatureStatesConfigMap(e.ObjectNew) &&
+				e.ObjectOld.GetResourceVersion() != e.ObjectNew.GetResourceVersion()
+		},
+		// Delete is not expected to occur
```
[the remark this answers is not in the document]
Comment (peterochodo) [reply addons/controllers/csi/vspherecsiconfig_controller.go:100]: deletion prevention of the internal-feature-states ConfigMap is a concern of the supervisor cluster. Tanzu Addons Manager simply observes that ConfigMap. If it is ever deleted, Tanzu Addons Manager will ignore that event and work with the last known settings.

File: addons/test/testutil/test_helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vijaykatam) [reply addons/test/testutil/test_helpers.go:165]: Is this method same as CreateResources with a check to get resource?

File: addons/controllers/csi/vspherecsiconfig_utils.go [hunk 4]
```diff
@@ -43,17 +43,31 @@ func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValues(ctx context
...
+	dvs.VSpherePVCSI.SupervisorMasterEndpointHostname = DefaultSupervisorMasterEndpointHostname
+	dvs.VSpherePVCSI.SupervisorMasterPort = DefaultSupervisorMasterPort
+	dvs.VSpherePVCSI.FeatureStates = map[string]string{}
+	featureStatesCM := &v1.ConfigMap{}
+	key := types.NamespacedName{Namespace: VSphereCSIFeatureStateNamespace, Name: VSphereCSIFeatureStateConfigMapName}
+	if err := r.Get(ctx, key, featureStatesCM); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("Error reading configmap '%s/%s': %v", key.Namespace, key.Name, err)
+		}
+		dvs.VSpherePVCSI.FeatureStates = nil
```
Comment (vijaykatam) [inline addons/controllers/csi/vspherecsiconfig_utils.go:64]: nit: Don't have to set nil as ytt check for empty map is same as checking for nil

File: addons/controllers/csi/vspherecsiconfig_constants.go [hunk 5]
```diff
@@ -26,10 +26,18 @@ const (
...
-	VSphereCSIResizerTimeout        = "300s"
-	VSphereCSIMinDeploymentReplicas = 1
-	VSphereCSIMaxDeploymentReplicas = 3
+	VSphereCSINamespace                 = "kube-system"
+	VSphereCSIProvisionTimeout          = "300s"
+	VSphereCSIAttachTimeout             = "300s"
+	VSphereCSIResizerTimeout            = "300s"
+	VSphereCSIMinDeploymentReplicas     = 1
+	VSphereCSIMaxDeploymentReplicas     = 3
+	VSphereCSIFeatureStateNamespace     = VSphereSystemCSINamepace
```
Comment (vijaykatam) [inline addons/controllers/csi/vspherecsiconfig_constants.go:35]: question: why define another constant pointing to "vmware-system-csi"?

**Review by vijaykatam — COMMENTED** [review]
Changes look fine to me. Please fill out PR template (add fixes issue and release note for tracking)before this PR can be merged.

**Comment by vijaykatam** [comment]
@peterochodo I filled out the PR template for you. Please follow this in the future if PR needs to be merged.

**Review by vijaykatam — APPROVED** [review]

**Pull request merged**